File: app/src/main/java/com/radardeal/app/ui/watches/WatchesScreen.kt
package com.radardeal.app.ui.watches

import androidx.compose.foundation.background
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.rounded.Add
import androidx.compose.material.icons.rounded.Delete
import androidx.compose.material.icons.rounded.Edit
import androidx.compose.material.icons.rounded.PauseCircle
import androidx.compose.material.icons.rounded.PlayCircle
import androidx.compose.material.icons.rounded.Refresh
import androidx.compose.material.icons.rounded.Visibility
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExtendedFloatingActionButton
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.SnackbarHost
import androidx.compose.material3.SnackbarHostState
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.radardeal.app.core.Formatters
import com.radardeal.app.data.repository.WatchWithStats
import com.radardeal.app.domain.model.ScanStatus
import com.radardeal.app.ui.components.BadgePill
import com.radardeal.app.ui.components.EmptyState
import com.radardeal.app.ui.components.LiveDot
import com.radardeal.app.ui.components.PrimaryButton
import com.radardeal.app.ui.components.RdCard
import com.radardeal.app.ui.components.SecondaryButton
import com.radardeal.app.ui.components.StatusBanner
import com.radardeal.app.ui.radarViewModel
import com.radardeal.app.ui.theme.LocalSpacing
import com.radardeal.app.ui.theme.RadarColors

/** The list of saved searches, with the per-watch actions from the spec. */
@Composable
fun WatchesScreen(
    onCreateWatch: () -> Unit,
    onEditWatch: (Long) -> Unit,
    onOpenLogin: () -> Unit,
) {
    val spacing = LocalSpacing.current
    val viewModel = radarViewModel { graph ->
        WatchesViewModel(
            appContext = graph.appContext,
            watchRepository = graph.watchRepository,
            coordinator = graph.coordinator,
            settingsStore = graph.settingsStore,
        )
    }
    val state by viewModel.state.collectAsStateWithLifecycle()
    val message by viewModel.message.collectAsStateWithLifecycle()
    val snackbarHost = remember { SnackbarHostState() }
    var pendingDeletion by remember { mutableStateOf<WatchWithStats?>(null) }

    LaunchedEffect(message) {
        val text = message ?: return@LaunchedEffect
        snackbarHost.showSnackbar(text)
        viewModel.consumeMessage()
    }

    Scaffold(
        containerColor = RadarColors.Background,
        snackbarHost = { SnackbarHost(snackbarHost) },
        floatingActionButton = {
            if (state.watches.isNotEmpty()) {
                ExtendedFloatingActionButton(
                    onClick = onCreateWatch,
                    containerColor = RadarColors.Accent,
                    contentColor = Color(0xFF00251E),
                    icon = { Icon(Icons.Rounded.Add, contentDescription = null) },
                    text = { Text("Nouvelle veille", style = MaterialTheme.typography.labelLarge) },
                )
            }
        },
    ) { padding ->
        LazyColumn(
            modifier = Modifier
                .fillMaxSize()
                .background(RadarColors.Background)
                .padding(padding),
            contentPadding = PaddingValues(
                start = spacing.screen,
                end = spacing.screen,
                bottom = 96.dp,
            ),
            verticalArrangement = Arrangement.spacedBy(spacing.md),
        ) {
            item(key = "title") {
                Column(
                    modifier = Modifier
                        .statusBarsPadding()
                        .padding(top = spacing.xl, bottom = spacing.sm),
                ) {
                    Text(
                        text = "Veilles",
                        style = MaterialTheme.typography.displayMedium,
                        color = RadarColors.TextPrimary,
                    )
                    Text(
                        text = when (state.watches.size) {
                            0 -> "Aucune recherche enregistrée"
                            1 -> "1 recherche enregistrée"
                            else -> "${state.watches.size} recherches enregistrées"
                        },
                        style = MaterialTheme.typography.bodyMedium,
                        color = RadarColors.TextSecondary,
                    )
                }
            }

            if (state.watches.isEmpty() && !state.isLoading) {
                item(key = "empty") {
                    EmptyState(
                        icon = Icons.Rounded.Visibility,
                        title = "Aucune veille",
                        message = "Crée ton premier radar : une marque, un modèle, un prix maximum — " +
                            "ou colle directement une recherche Vinted.",
                        action = { PrimaryButton(text = "Créer un radar", onClick = onCreateWatch) },
                    )
                }
            }

            items(state.watches, key = { it.watch.id }) { entry ->
                WatchCard(
                    entry = entry,
                    scanning = entry.watch.id in state.scanningIds,
                    onToggleActive = { viewModel.setActive(entry.watch.id, !entry.watch.isActive) },
                    onScanNow = { viewModel.scanNow(entry.watch.id) },
                    onEdit = { onEditWatch(entry.watch.id) },
                    onDelete = { pendingDeletion = entry },
                    onOpenLogin = onOpenLogin,
                )
            }
        }
    }

    pendingDeletion?.let { target ->
        AlertDialog(
            onDismissRequest = { pendingDeletion = null },
            containerColor = RadarColors.SurfaceElevated,
            titleContentColor = RadarColors.TextPrimary,
            textContentColor = RadarColors.TextSecondary,
            title = { Text("Supprimer cette veille ?") },
            text = {
                Text(
                    "« ${target.watch.name} » et les ${target.knownCount} annonces qu'elle a " +
                        "collectées seront supprimées définitivement."
                )
            },
            confirmButton = {
                TextButton(onClick = {
                    viewModel.delete(target.watch.id)
                    pendingDeletion = null
                }) {
                    Text("Supprimer", color = RadarColors.Danger)
                }
            },
            dismissButton = {
                TextButton(onClick = { pendingDeletion = null }) {
                    Text("Annuler", color = RadarColors.TextSecondary)
                }
            },
        )
    }
}

@Composable
private fun WatchCard(
    entry: WatchWithStats,
    scanning: Boolean,
    onToggleActive: () -> Unit,
    onScanNow: () -> Unit,
    onEdit: () -> Unit,
    onDelete: () -> Unit,
    onOpenLogin: () -> Unit,
) {
    val spacing = LocalSpacing.current
    val watch = entry.watch

    RdCard(modifier = Modifier.fillMaxWidth()) {
        Column(
            modifier = Modifier.padding(spacing.lg),
            verticalArrangement = Arrangement.spacedBy(spacing.md),
        ) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(spacing.md),
                verticalAlignment = Alignment.Top,
            ) {
                Column(modifier = Modifier.weight(1f)) {
                    Text(
                        text = watch.name,
                        style = MaterialTheme.typography.headlineMedium,
                        color = RadarColors.TextPrimary,
                        maxLines = 2,
                        overflow = TextOverflow.Ellipsis,
                    )
                    Text(
                        text = watch.criteriaSummary(),
                        style = MaterialTheme.typography.bodyMedium,
                        color = RadarColors.TextSecondary,
                        maxLines = 2,
                        overflow = TextOverflow.Ellipsis,
                    )
                }

                Row(
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(6.dp),
                ) {
                    LiveDot(active = watch.isActive)
                    Text(
                        text = if (watch.isActive) "LIVE" else "PAUSE",
                        style = MaterialTheme.typography.labelMedium,
                        color = if (watch.isActive) RadarColors.Live else RadarColors.TextTertiary,
                    )
                }
            }

            Row(horizontalArrangement = Arrangement.spacedBy(spacing.sm)) {
                if (watch.isUltra) {
                    BadgePill(text = "⚡ ULTRA")
                }
                BadgePill(
                    text = if (watch.isUltra) {
                        watch.frequency.description
                    } else {
                        Formatters.duration(watch.intervalSeconds)
                    },
                    contentColor = if (watch.isUltra) RadarColors.Accent else RadarColors.TextSecondary,
                    background = RadarColors.SurfaceElevated,
                )
                BadgePill(
                    text = "${entry.knownCount} ANNONCES",
                    contentColor = RadarColors.TextSecondary,
                    background = RadarColors.SurfaceElevated,
                )
                if (entry.newCount > 0) {
                    BadgePill(text = "${entry.newCount} NOUVEAU")
                }
            }

            Text(
                text = "Dernier scan : ${Formatters.relativeTime(watch.lastScanAt)}" +
                    if (watch.lastScanStatus == ScanStatus.NEVER) "" else " · ${watch.lastScanStatus.label}",
                style = MaterialTheme.typography.labelSmall,
                color = RadarColors.TextTertiary,
            )

            if (watch.lastScanStatus.isProblem) {
                StatusBanner(
                    status = watch.lastScanStatus,
                    action = if (watch.lastScanStatus == ScanStatus.NEEDS_LOGIN) {
                        {
                            SecondaryButton(
                                text = "Se connecter à Vinted",
                                onClick = onOpenLogin,
                                contentColor = RadarColors.Accent,
                            )
                        }
                    } else {
                        null
                    },
                )
            }

            // The four per-watch actions. The row scrolls horizontally so that a narrow
            // screen or a large font scale never clips "Supprimer" off the card.
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .horizontalScroll(rememberScrollState()),
                horizontalArrangement = Arrangement.spacedBy(spacing.sm),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                SecondaryButton(
                    text = if (watch.isActive) "Pause" else "Activer",
                    onClick = onToggleActive,
                    icon = if (watch.isActive) Icons.Rounded.PauseCircle else Icons.Rounded.PlayCircle,
                    contentColor = RadarColors.TextPrimary,
                )
                if (scanning) {
                    Box(
                        modifier = Modifier.size(48.dp),
                        contentAlignment = Alignment.Center,
                    ) {
                        CircularProgressIndicator(
                            modifier = Modifier.size(20.dp),
                            strokeWidth = 2.dp,
                            color = RadarColors.Accent,
                        )
                    }
                } else {
                    SecondaryButton(
                        text = "Scanner",
                        onClick = onScanNow,
                        icon = Icons.Rounded.Refresh,
                        contentColor = RadarColors.Accent,
                    )
                }
                SecondaryButton(
                    text = "Modifier",
                    onClick = onEdit,
                    icon = Icons.Rounded.Edit,
                    contentColor = RadarColors.TextPrimary,
                )
                SecondaryButton(
                    text = "Supprimer",
                    onClick = onDelete,
                    icon = Icons.Rounded.Delete,
                    contentColor = RadarColors.Danger,
                )
            }
        }
    }
}

File: ios/App/RadarDeal/Screens/WatchesScreen.swift
import SwiftUI
import RadarDealCore

/// The list of saved searches, with the per-watch actions.
struct WatchesScreen: View {
    @Environment(AppGraph.self) private var graph
    @Environment(RadarCoordinator.self) private var coordinator

    @State private var watches: [Watch] = []
    @State private var counts: [Int64: Int] = [:]
    @State private var editing: Watch?
    @State private var creating = false
    @State private var pendingDeletion: Watch?
    @State private var message: String?
    @State private var scanningID: Int64?

    var body: some View {
        NavigationStack {
            ScrollView {
                LazyVStack(alignment: .leading, spacing: RD.Space.md) {
                    VStack(alignment: .leading, spacing: 2) {
                        Text("Veilles").font(.rdDisplayMedium).foregroundStyle(RD.textPrimary)
                        Text(subtitle).font(.rdBodySmall).foregroundStyle(RD.textSecondary)
                    }
                    .padding(.top, RD.Space.xl)

                    if watches.isEmpty {
                        EmptyStateView(
                            systemImage: "eye",
                            title: "Aucune veille",
                            message: "Crée ton premier radar : une marque, un modèle, un prix "
                                + "maximum — ou colle directement une recherche Vinted."
                        ) {
                            PrimaryButton(title: "Créer un radar") { creating = true }
                        }
                    }

                    ForEach(watches) { watch in
                        watchCard(watch)
                    }
                }
                .padding(.horizontal, RD.Space.screen)
                .padding(.bottom, 96)
            }
            .background(RD.background)
            .scrollContentBackground(.hidden)
            .overlay(alignment: .bottomTrailing) {
                if !watches.isEmpty {
                    Button { creating = true } label: {
                        Label("Nouvelle veille", systemImage: "plus")
                            .font(.rdTitle)
                            .foregroundStyle(RD.onAccent)
                            .padding(.horizontal, 20)
                            .frame(height: 56)
                            .background(RD.accent, in: Capsule())
                    }
                    .buttonStyle(.plain)
                    .padding(RD.Space.screen)
                }
            }
            .sheet(isPresented: $creating) {
                WatchEditorScreen(existing: nil) { reload() }
            }
            .sheet(item: $editing) { watch in
                WatchEditorScreen(existing: watch) { reload() }
            }
            .alert("Supprimer cette veille ?", isPresented: .constant(pendingDeletion != nil)) {
                Button("Annuler", role: .cancel) { pendingDeletion = nil }
                Button("Supprimer", role: .destructive) {
                    if let target = pendingDeletion {
                        graph.store.deleteWatch(id: target.id)
                        coordinator.forget(target.id)
                        reload()
                    }
                    pendingDeletion = nil
                }
            } message: {
                if let target = pendingDeletion {
                    Text("« \(target.name) » et les \(counts[target.id] ?? 0) annonces qu'elle "
                         + "a collectées seront supprimées définitivement.")
                }
            }
            .overlay(alignment: .bottom) {
                if let message {
                    Text(message)
                        .font(.rdBodySmall)
                        .foregroundStyle(RD.textPrimary)
                        .padding(RD.Space.md)
                        .background(RD.surfaceElevated, in: Capsule())
                        .padding(.bottom, 80)
                        .task {
                            try? await Task.sleep(for: .seconds(3))
                            self.message = nil
                        }
                }
            }
        }
        .task { reload() }
        .onChange(of: coordinator.lastTickAt) { _, _ in reload() }
    }

    private var subtitle: String {
        switch watches.count {
        case 0: return "Aucune recherche enregistrée"
        case 1: return "1 recherche enregistrée"
        default: return "\(watches.count) recherches enregistrées"
        }
    }

    private func watchCard(_ watch: Watch) -> some View {
        RDCard {
            VStack(alignment: .leading, spacing: RD.Space.md) {
                HStack(alignment: .top) {
                    VStack(alignment: .leading, spacing: 2) {
                        Text(watch.name)
                            .font(.rdHeadline)
                            .foregroundStyle(RD.textPrimary)
                            .lineLimit(2)
                        Text(watch.criteriaSummary())
                            .font(.rdBodySmall)
                            .foregroundStyle(RD.textSecondary)
                            .lineLimit(2)
                    }
                    Spacer(minLength: RD.Space.sm)
                    HStack(spacing: 6) {
                        LiveDot(active: watch.isActive)
                        Text(watch.isActive ? "LIVE" : "PAUSE")
                            .font(.rdLabel)
                            .foregroundStyle(watch.isActive ? RD.live : RD.textTertiary)
                    }
                }

                HStack(spacing: RD.Space.sm) {
                    if watch.isUltra { BadgePill(text: "⚡ ULTRA") }
                    BadgePill(
                        text: watch.isUltra
                            ? watch.frequency.descriptionText
                            : Formatters.duration(seconds: watch.intervalSeconds),
                        foreground: watch.isUltra ? RD.accent : RD.textSecondary,
                        background: RD.surfaceElevated
                    )
                    BadgePill(text: "\(counts[watch.id] ?? 0) ANNONCES",
                              foreground: RD.textSecondary, background: RD.surfaceElevated)
                }

                Text("Dernier scan : \(Formatters.relativeTime(watch.lastScanAt))"
                     + (watch.lastScanStatus == .never ? "" : " · \(watch.lastScanStatus.label)"))
                    .font(.rdLabelSmall)
                    .foregroundStyle(RD.textTertiary)

                if watch.lastScanStatus.isProblem {
                    StatusBanner(status: watch.lastScanStatus)
                }

                ScrollView(.horizontal, showsIndicators: false) {
                    HStack(spacing: RD.Space.sm) {
                        SecondaryButton(
                            title: watch.isActive ? "Pause" : "Activer",
                            systemImage: watch.isActive ? "pause.circle" : "play.circle"
                        ) {
                            graph.store.setActive(id: watch.id, active: !watch.isActive)
                            if !watch.isActive { coordinator.markDueNow(watch.id) }
                            reload()
                        }

                        if scanningID == watch.id {
                            ProgressView().tint(RD.accent).frame(width: 48, height: 48)
                        } else {
                            SecondaryButton(title: "Scanner", systemImage: "arrow.clockwise",
                                            tint: RD.accent) {
                                Task { await scanNow(watch) }
                            }
                        }

                        SecondaryButton(title: "Modifier", systemImage: "pencil") {
                            editing = watch
                        }
                        SecondaryButton(title: "Supprimer", systemImage: "trash",
                                        tint: RD.danger) {
                            pendingDeletion = watch
                        }
                    }
                }
            }
            .padding(RD.Space.lg)
        }
    }

    private func scanNow(_ watch: Watch) async {
        scanningID = watch.id
        defer { scanningID = nil }

        let report = await coordinator.scanNow(watchID: watch.id)
        reload()

        guard let report else { message = "Cette veille est introuvable."; return }
        if report.status.isProblem {
            message = report.status.label
        } else if report.wasBaseline {
            message = "Référence enregistrée : \(report.parsedCount) annonces."
        } else if report.newCount > 0 {
            message = "\(report.newCount) nouvelle\(report.newCount > 1 ? "s" : "") annonce"
                + (report.newCount > 1 ? "s" : "")
        } else if report.parsedCount == 0 {
            message = "Aucun résultat pour cette recherche."
        } else {
            message = "Aucune nouveauté (\(report.parsedCount) annonces analysées)."
        }
    }

    private func reload() {
        watches = graph.store.allWatches()
        counts = Dictionary(
            watches.map { ($0.id, graph.store.listings(watchID: $0.id).count) },
            uniquingKeysWith: { a, _ in a }
        )
    }
}
